### PR TITLE
Preferences - Allow screen to sleep

### PR DESCRIPTION
**Pull request by smartyw**
This change adds a new setting in the Android application which allows the user to specify that the screen should be allowed to sleep after a period of inactivity, per the Android Settings->Display->Sleep configuration. The change relates to Issue 
 196 (https://github.com/organicmaps/organicmaps/issues/196)

### CODE REVIEW

**Comment by vng** [comment]
- Better to remove merge commit and **rebase** your commit on top of master.
- Sign-off your commit
- @rtsisyk are we ok with english only localization here?

File: android/res/values/strings.xml [hunk 1]
```diff
@@ -1128,6 +1128,8 @@
 	<string name="power_managment_setting_never">Never</string>
 	<string name="power_managment_setting_auto">Automatic</string>
 	<string name="power_managment_setting_manual_max">Maximum power saving</string>
+	<string name="enable_screen_sleep">Allow screen to sleep</string>
```
Comment (rtsisyk) [inline android/res/values/strings.xml:1131]: Please update data/strings.txt and run `./tools/unix/generate_locationzation.sh` instead.

**Review by rtsisyk — COMMENTED** [review]
Hi @smartyw,

Thank you for your contribution! 

I don't see any serious problems with the code itself. However, I'm not fully convinced that UX side is good for users. Please give me a couple of days to test this feature in a device and understand how it works in real conditions.

File: android/res/values/strings.xml (hunk 1 above)
Comment (smartyw) [reply android/res/values/strings.xml:1131]: Assuming you mean data/strings/strings.txt and that you're OK with Google Translate generated values? Re: UX please review https://github.com/organicmaps/organicmaps/issues/196 which I hope gives a good explanation. It's critical that complete control over battery use is available on longer bike or hiking trips where you don't have regular (daily) access to electricity. I've tested this a few times and the application is impractical to use if you have to keep manually switching the screen off. See https://the-bikepacker.blogspot.com/search/label/Trips for some examples :-)

Comment (smartyw) [reply android/res/values/strings.xml:1131]: p.s. I use a Bluetooth paired watch as a trusted device and this stops the screen from actually locking. That way after the display sleeps, if I need to see the map again, I just tap the screen to wake it up. Easy to do when cycling.

**Comment by smartyw** [comment]
You save enormous amounts of battery because the screen is a major consumer of power. You don't need to know your position all the time when cycling long distances so it's something you check occasionally only and in so doing save lots of power. To be clear... you do need a turned on screen, just not all the time and in fact only a small proportion of the time.

On my touring bike with heavy camping gear I only average about 10 miles an hour so cycling for 6 to 8 hours is typical. If the screen was on for the entire time, battery would be significantly depleted by the end. With the screen only on when required, I can get to my destination and still have a battery level of 90% or more. And factor into that, the fact that sometimes trips span multiple days and with no opportunity to charge. I agree this is not a common requirement, but I'll bet it's also less rare than you might imagine. A checkbox defaulting to OFF seems a minor thing which would not inconvenience other users at all but would be a tremendous benefit to travelers like myself.

I'm about to start updating the strings.txt file as requested. If you are thinking of rejecting the feature as an "exceptional scenario" could you please tell me as soon as possible? Thanks.

File: android/res/values/strings.xml (hunk 1 above)
Comment (smartyw) [reply android/res/values/strings.xml:1131]: @rtsisyk Hi, any tips for setting up Ruby to run the localization script? I installed using brew and moved the new version to the front of my PATH (I'm on a Mac). When I run the script I get 

```
<internal:/usr/local/Cellar/ruby/3.0.1/lib/ruby/3.0.0/rubygems/core_ext/kernel_require.rb>:85:in `require': cannot load such file -- safe_yaml/load (LoadError)
	from <internal:/usr/local/Cellar/ruby/3.0.1/lib/ruby/3.0.0/rubygems/core_ext/kernel_require.rb>:85:in `require'
	from /Users/martin/AndroidStudioProjects/organicmaps/tools/twine/lib/twine/plugin.rb:2:in `<top (required)>'
	from <internal:/usr/local/Cellar/ruby/3.0.1/lib/ruby/3.0.0/rubygems/core_ext/kernel_require.rb>:85:in `require'
	from <internal:/usr/local/Cellar/ruby/3.0.1/lib/ruby/3.0.0/rubygems/core_ext/kernel_require.rb>:85:in `require'
	from /Users/martin/AndroidStudioProjects/organicmaps/tools/twine/lib/twine.rb:25:in `<module:Twine>'
	from /Users/martin/AndroidStudioProjects/organicmaps/tools/twine/lib/twine.rb:1:in `<top (required)>'
	from <internal:/usr/local/Cellar/ruby/3.0.1/lib/ruby/3.0.0/rubygems/core_ext/kernel_require.rb>:85:in `require'
	from <internal:/usr/local/Cellar/ruby/3.0.1/lib/ruby/3.0.0/rubygems/core_ext/kernel_require.rb>:85:in `require'
	from ./tools/unix/../../tools/twine/bin/twine:2:in `<main>'
```
and
```
martin@Raticate organicmaps % which ruby
/usr/local/opt/ruby/bin/ruby
martin@Raticate organicmaps % ruby --version
ruby 3.0.1p64 (2021-04-05 revision 0fb782ee38) [x86_64-darwin20]
```
Thanks

**Comment by biodranik** [comment]
Saving battery is great, we're strongly supporting that, so don't worry about rejection :) I remember using MapsMe in New Zealand great walk for 3 days without charging, so the importance is very clear.

The questions are only about the best possible experience/implementation for users. So far I see a number of theoretical options:

1. Current implementation (manually lock the device)
2. Dim/turn off the screen after some inactivity, but continue to detect location (in theory some power savings, but how good are they, if location and renderer still consume CPU?)
3. Autolock the device (great power savings but location detection may be delayed after unlocking the device)
4. Allow users to disable location in the app and this in turn will activate automatic screen lock (may be useful in some cases)

May be there are other options. What do you think?

**Comment by smartyw** [comment]
But I like to worry ;-)

Let me think about the options but to be clear (1) does not manually lock the device. It allows the operating system display sleep setting to take effect. Right now the app disables this. So, if the user wants the screen on even after as much as say 30 minutes of inactivity, they can do.... by choosing that value in the Android settings screen. For me, I have the screen sleep time set to 30 seconds which allows me to tap/swipe my phone to wake up the screen.... see my current location (which might be on the planned route or maybe I've deviated) and then I carry on cycling and the screen sleeps automatically soon after. So, apart from choosing this behaviour manually, after that everything is automatic.

I'm not sure what impact on power or user experience switching off location would have. I'm not sure how GPS works but if it's like Bluetooth scanning, with a scan interval (how often to listen) and a scan window (how long to listen for) then maybe you could keep location on when the screen sleeps but reduce the interval and window size to save power?

I'll see if I can get the strings.txt file updated in the meantime.... having some Ruby problems but that's a whole other story :-)

Cheers

**Comment by biodranik** [comment]
My point is that if your location update does not stop when the screen is off (e.g. the device is not locked, only the screen is turned off), then map renderer will also work and redraw on every map update. It means that battery savings may be negligible, so it would be great to test.

Btw, I've never heard before about turning the screen off without locking the device/pausing all applications (except those specifically designed to work in the background, like voice navigation). @rtsisyk is it a common behavior?

File: android/res/values/strings.xml (hunk 1 above)
Comment (biodranik) [reply android/res/values/strings.xml:1131]: There are some workarounds in google, I don't remember which one helped me.

**Comment by smartyw** [comment]
OK, I see what you mean. When I was manually showing the home screen to allow the OS display sleep to take effect, leaving maps.me in the background, my experience was very good. I have the impression that GPS location tracking was suspended too because I remember waking up the screen to see the map and having to wait whilst my location was determined.

This is worth testing of course, I totally agree. If I can find the time I'll see what I can do.

On letting the screen sleep but keeping the device unlocked, see https://helpdeskgeek.com/how-to/how-to-use-android-trusted-devices/

File: android/res/values/strings.xml (hunk 1 above)
Comment (smartyw) [reply android/res/values/strings.xml:1131]: OK, I'll hit Google. Never worked with Ruby before.

**Comment by biodranik** [comment]
Thanks for the link! From the description, it looks like apps are paused, and only unlock procedure is bypassed. I assume that GPS will be stopped too in that scenario. Interesting.
Without a trusted device, the android will lock itself after a timeout, right?

**Comment by smartyw** [comment]
Yes, I believe that is the case. The trusted device plus optional screen sleeping behaviour controlled by a setting is a killer combination for those of us wanting to squeeze as much battery life out of their phone as possible :-)

**Comment by smartyw** [comment]
Sorry, hadn't intended to commit the change to .gitignore.

**Comment by smartyw** [comment]
Sorry, I hadn't intended to commit the change to android/.gitignore.

**Comment by smartyw** [comment]
Do you guys need me to do anything? I realise I made a mistake commiting .gitignore but I'm assuming you can deal with that? Sorry, but I do all my development work alone so I'm not as familiar with all things git as I should be.

**Comment by biodranik** [comment]
No worries, we'll take care of the code. Just need some free time for that )

**Comment by smartyw** [comment]
Thank you, much appreciated. I can relate :-)

**Comment by smartyw** [comment]
FYI I'm planning a 3 day trip at the beginning of next month. It would be tremendous if this change were in the application and the full store release before then so I can benefit from the improvement. I understand if this is not possible but would be delighted if somehow it could be. All the best.

**Comment by rtsisyk** [comment]
Hi Martin, sorry for the delay. 

I don't see any reason why we can't push this good patch. Re-based and pushed as 470bc6f231bcb93a8024e8407d5b7903affdd9d2
A new beta will be available in an hour from https://appdistribution.firebase.google.com/pub/i/4fa64e2439ca20ff.

Thank you for the contribution!

**Comment by smartyw** [comment]
That's wonderful, thank you so much!

**Comment by keunes** [comment]
Thanks @smartyw for this change and the team for your overall work on the app! :)

For me it's not even about saving battery, but about being able to put the phone in my pocket whilst cycling and still hearing instructions over the headphone. Hope (but I believe) this scenario will work, too, with this update.

**Comment by smartyw** [comment]
@keunes Glad you like it and that it has other benefits :-)

This is a brilliant application and quite possibly my favourite open source project. The core team deserve all credit and our thanks for bringing it back into the world of open source fully.

My blog post on navigation has been updated to recommend Organic Maps.

https://the-bikepacker.blogspot.com/2020/08/navigation.html

**Comment by keunes** [comment]
Nice overview you have there in that blogpost. Looking into the Android options, I figured that 'pinning' the screen can be an alternative to 'smart lock' (which might not work for everybody). It also displays the relevant app when activating the screen, without the need to enter a pin.

**Comment by rtsisyk** [comment]
You can also try https://maps.omaps.app/ for planning routes. Not integrated with app, but you can download KMLs and import into app.

**Comment by smartyw** [comment]
Thanks, @rtsisyk . Another good tip.

**Comment by pastk** [comment]
@smarty could you please clarify your bike touring use case - do you use OM in navigation mode (when it shows ETA, next turn etc.) or just in a map browsing mode?

**Comment by smartyw** [comment]
@pastk I rarely use navigation (turn by turn) mode. Mostly I save battery power by having my screen asleep all the time unless I need to check my position or how far to the next turn. Then I press the power button on the side of my phone to wake the screen up and I then see the OM map which refreshes the current location fairly quickly usually. My phone is usually mounted in a waterproof bag on the handlebars of my bike FYI.

**Pull request closed without merging**